### PR TITLE
feat: add pre-build evaluation metrics

### PR DESCRIPTION
**Pull request by MaksymAI**
## Example of run for answer correctness

```
Question: What powers the sun and what is its primary function?
Answer: The sun is powered by nuclear fission, similar to nuclear reactors on Earth. Its primary function is to provide light to the solar system.
Ground Truth: The sun is powered by nuclear fusion, where hydrogen atoms fuse to form helium. This fusion process releases a tremendous amount of energy. The sun provides heat and light, which are essential for life on Earth.
Classifications:
{
  "TP": [
    "Its primary function is to provide light to the solar system."
  ],
  "FP": [
    "The sun is powered by nuclear fission, similar to nuclear reactors on Earth."
  ],
  "FN": [
    "The sun is powered by nuclear fusion, where hydrogen atoms fuse to form helium.",
    "This fusion process releases a tremendous amount of energy.",
    "The sun provides heat and light, which are essential for life on Earth."
  ]
}
F1 Score: 0.3333333333333333
Similarity Score: 0.7
Final Score: 0.425
--------------------------------------------------
Question: What is the boiling point of water?
Answer: The boiling point of water is 100 degrees Celsius at sea level.
Ground Truth: The boiling point of water is 100 degrees Celsius (212 degrees Fahrenheit) at sea level. The boiling point can change with altitude.
Classifications:
{
  "TP": [
    "The boiling point of water is 100 degrees Celsius at sea level."
  ],
  "FP": [],
  "FN": [
    "The boiling point of water is 100 degrees Celsius (212 degrees Fahrenheit) at sea level.",
    "The boiling point can change with altitude."
  ]
}
F1 Score: 0.5
Similarity Score: 1.0
Final Score: 0.625
--------------------------------------------------
Answer Correctness Scores:
[0.425, 0.625]

```

### CODE REVIEW

File: dynamiq/evaluations/metrics/factual_correctness.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+                logger.debug(f"Reference Claims: {reference_claims}")
+                logger.debug(f"TP: {tp}, FP: {fp}, FN: {fn}")
+                logger.debug(f"Score: {score}")
+                logger.debug("-" * 50)
+
+        return final_scores
+
+
+# Example usage
+if __name__ == "__main__":
```
Comment (vitalii-dynamiq) [inline dynamiq/evaluations/metrics/factual_correctness.py:208]: @MaksymAI should we move such logic into examples?

File: dynamiq/evaluations/metrics/faithfulness.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
+import logging
+
+from dynamiq.components.evaluators.llm_evaluator import LLMEvaluator
+from dynamiq.nodes.llms import BaseLLM, OpenAI
+
+# Configure logging
+logger = logging.getLogger(__name__)
+
+
+class FaithfulnessEvaluator:
```
Comment (vitalii-dynamiq) [inline dynamiq/evaluations/metrics/faithfulness.py:10]: should we inherit from LLMEvaluator?

File: dynamiq/evaluations/metrics/answer_correctness.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
+import json
```
Comment (vitalii-dynamiq) [inline dynamiq/evaluations/metrics/answer_correctness.py:1]: @delamainer would these be supported on the backend now?

File: dynamiq/evaluations/metrics/answer_correctness.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+from dynamiq.components.evaluators.llm_evaluator import LLMEvaluator
+from dynamiq.nodes.llms import BaseLLM, OpenAI
+
+# Configure logging
+logger = logging.getLogger(__name__)
+
+
+class AnswerCorrectnessEvaluator:
+    def __init__(self, llm: BaseLLM, weights: list[float] = [0.75, 0.25]):
```
Comment (acoola) [inline dynamiq/evaluations/metrics/answer_correctness.py:12]: Let's use Pydantic models to have auto validation

File: dynamiq/evaluations/metrics/context_precision.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+            return 0.0
+        for i, verdict in enumerate(verdicts):
+            if verdict == 1:
+                cumulative_hits += 1
+                precision_at_i = cumulative_hits / (i + 1)
+                numerator += precision_at_i
+        average_precision = numerator / total_relevant
+        return average_precision
+
+    def evaluate(
```
Comment (acoola) [inline dynamiq/evaluations/metrics/context_precision.py:141]: Everywhere we used `run`

File: dynamiq/evaluations/metrics/faithfulness.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+                        ]
+                    },
+                },
+            ],
+            llm=self.llm,
+        )
+
+    def evaluate(
+        self,
+        questions: list[str],
```
Comment (acoola) [inline dynamiq/evaluations/metrics/faithfulness.py:145]: Please check if it is possible to unify evaluate/run parameters and their order. I some I see `contexts`, in other `contexts_list`.

File: dynamiq/evaluations/metrics/answer_correctness.py (hunk 3 above)
Comment (MaksymAI) [reply dynamiq/evaluations/metrics/answer_correctness.py:1]: Probably not these RAGAS-like complex metrics yet, but we will add their support to catalyst as soon as we merge them into dynamiq

File: dynamiq/evaluations/metrics/answer_correctness.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+        # Initialize the LLMEvaluators and store them as class attributes
+
+        # Extract Statements Evaluator
+        extract_instructions = (
+            "For each input text, extract the key statements.\n"
+            "- Keep statements concise and focused.\n"
+            "- Return the statements as a JSON array of strings.\n"
+            "- Ensure that your response is valid JSON, using double quotes for all strings."
+        )
+        self.statement_extractor = LLMEvaluator(
```
[the remark this answers is not in the document]
Comment (MaksymAI) [reply dynamiq/evaluations/metrics/answer_correctness.py]: It's the RAGAS metrics implementation rewritten with our framework.

Some metrics could be simplified and written with just one prompt (like returning 1 or 0), but to have meaningful scores in the range from 0 to 1 is a complicated task for LLMs. That's why the RAGAS split the task into subtasks, such as extracting statements, classifying them separately, and then calculating a classification metric like F1 based on the classified statements

File: dynamiq/evaluations/metrics/faithfulness.py (hunk 6 above)
Comment (MaksymAI) [reply dynamiq/evaluations/metrics/faithfulness.py:335]: I tried to unify the names and order of the parameters.

However, the exact parameters will differ for each metric. Some metrics require ground truth, while others don't need it but instead require context, for instance.

File: dynamiq/evaluations/metrics/context_precision.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+            return 0.0
+        for i, verdict in enumerate(verdicts):
+            if verdict == 1:
+                cumulative_hits += 1
+                precision_at_i = cumulative_hits / (i + 1)
+                numerator += precision_at_i
+        average_precision = numerator / total_relevant
+        return average_precision
+
+    def run(
```
Comment (acoola) [inline dynamiq/evaluations/metrics/context_precision.py:219]: With schema input I mean that `run` func require schema:
```
def run(self, input_data: ContextPrecisionInput) -> list[float]:
```

File: dynamiq/evaluations/metrics/context_precision.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+    """
+
+    llm: BaseLLM
+
+    # Private attribute (not part of the Pydantic model fields)
+    _context_precision_evaluator: LLMEvaluator = PrivateAttr()
+
+    class Config:
+        arbitrary_types_allowed = True  # Allow arbitrary types
+
```
Comment (acoola) [inline dynamiq/evaluations/metrics/context_precision.py:81]: I think we should add type for each evaluator (probably with some BaseEvaluator to limit duplication) in similar way as we define for Nodes:
https://github.com/dynamiq-ai/dynamiq/blob/main/dynamiq/nodes/node.py#L231

```
    @computed_field
    @cached_property
    def type(self) -> str:
        return f"{self.__module__.rsplit('.', 1)[0]}.{self.__class__.__name__}"
```
As well need to modify `__init__.py`
```
from .context_precision import ContextPrecisionEvaluator
```

As result we could get proper class by type:
```
from dynamiq.loaders.yaml import WorkflowYAMLLoader

metric_evaluator_cls = WorkflowYAMLLoader.get_entity_by_type("dynamiq.evaluations.metrics.ContextPrecisionEvaluator")
```

**Review by acoola — APPROVED** [review]
Great job

**Pull request merged**